File: voice/transcribe.py
#!/usr/bin/env python3
"""Transcribe one audio file to text — the model half of Alfred's voice notes.

A Discord voice message is an Ogg/Opus attachment. bot.js downloads it and hands
the path here; this prints the transcript to stdout and nothing else, so the JS
side can treat stdout as the text verbatim.

Two steps, both cheap on this box (measured ~0.7s inference + ~2s model load for
a short note, CPU-only):

  1. ffmpeg decodes Opus -> 16 kHz mono PCM. The model wants 16 kHz mono, and
     ffmpeg is the one decoder that reliably reads Discord's Opus.
  2. NVIDIA Parakeet-TDT-0.6B, int8 ONNX, via onnx-asr on onnxruntime. Loaded
     from the local Hugging Face cache with downloads disabled — the model is
     pre-fetched by scripts/setup-voice.sh, and a live turn must never block on
     the network. Set ALFRED_VOICE_ALLOW_DOWNLOAD=1 to let the first run fetch it.

Contract with the caller (voice/transcribe.js):
  argv[1]         path to the audio file
  stdout          the transcript (may be empty on silence)
  exit 0          success (empty stdout = nothing intelligible heard)
  exit 2          dependencies/model missing or a decode/inference error
The exit split lets the bridge tell "couldn't run" from "ran, heard nothing".

Usage: python transcribe.py <audio_path>
"""
import os
import shutil
import subprocess
import sys
import tempfile

MODEL = "nemo-parakeet-tdt-0.6b-v2"


def eprint(*a):
    print(*a, file=sys.stderr, flush=True)


def find_ffmpeg():
    """ffmpeg next to this interpreter (the venv carries a static build) wins,
    then an explicit override, then PATH. bot.js spawns python directly without
    activating the venv, so the venv's bin is not on PATH — look there first."""
    override = os.environ.get("ALFRED_FFMPEG")
    if override:
        return override
    sibling = os.path.join(os.path.dirname(sys.executable), "ffmpeg")
    if os.path.isfile(sibling) and os.access(sibling, os.X_OK):
        return sibling
    return shutil.which("ffmpeg")


def decode_to_wav(src, ffmpeg, dst):
    # -vn drops any cover art; a single-stream voice note has none, but a
    # forwarded file might. 16 kHz mono s16 is what the model's preprocessor
    # expects and what onnx-asr documents.
    subprocess.run(
        [ffmpeg, "-y", "-loglevel", "error", "-vn",
         "-i", src, "-ar", "16000", "-ac", "1", "-f", "wav", dst],
        check=True,
    )


def main():
    if len(sys.argv) != 2:
        eprint("usage: transcribe.py <audio_path>")
        return 2
    src = sys.argv[1]
    if not os.path.isfile(src):
        eprint(f"no such file: {src}")
        return 2

    ffmpeg = find_ffmpeg()
    if not ffmpeg:
        eprint("ffmpeg not found (looked next to python, then PATH)")
        return 2

    # Downloads off by default: a live turn loads from cache or fails fast,
    # never stalls on the Hub. setup-voice.sh flips this to pre-fetch.
    if os.environ.get("ALFRED_VOICE_ALLOW_DOWNLOAD") != "1":
        os.environ.setdefault("HF_HUB_OFFLINE", "1")

    tmp = tempfile.NamedTemporaryFile(suffix=".wav", delete=False)
    tmp.close()
    try:
        try:
            decode_to_wav(src, ffmpeg, tmp.name)
        except subprocess.CalledProcessError as e:
            eprint(f"ffmpeg decode failed: {e}")
            return 2

        try:
            import onnx_asr
        except ImportError as e:
            eprint(f"onnx-asr not installed: {e} — run scripts/setup-voice.sh")
            return 2

        try:
            model = onnx_asr.load_model(MODEL, quantization="int8")
            text = model.recognize(tmp.name)
        except Exception as e:  # model missing from cache, onnxruntime error, …
            eprint(f"transcription failed: {type(e).__name__}: {e}")
            return 2

        sys.stdout.write((text or "").strip())
        sys.stdout.flush()
        return 0
    finally:
        try:
            os.unlink(tmp.name)
        except OSError:
            pass


if __name__ == "__main__":
    sys.exit(main())

File: voice/transcribe.js
// The Node side of voice notes: hand a downloaded audio file to the Python
// transcriber (voice/transcribe.py) and get its text back. The heavy lifting —
// ffmpeg decode + the Parakeet int8 ONNX model — lives in Python because that's
// where onnxruntime and the model are; this module is only the bridge, the same
// shape strength/plot.js draws over strength/plot.py.
//
// Spawned async (not spawnSync): a voice note takes ~3s to load the model and
// transcribe, and blocking the event loop that long would freeze the whole bot.

import { spawn, spawnSync } from "node:child_process";
import { existsSync } from "node:fs";
import path from "node:path";
import os from "node:os";
import { fileURLToPath } from "node:url";

const TRANSCRIBE_PY = fileURLToPath(new URL("./transcribe.py", import.meta.url));

// The interpreter to run the transcriber with. Prefer an explicit override, then
// the user's standard venv (which carries onnx-asr + a static ffmpeg), then
// whatever python3 is on PATH. Kept resolvable so transcriptionAvailable() and
// the tests can gate on it.
export function pythonBin() {
  if (process.env.VOICE_PYTHON) return process.env.VOICE_PYTHON;
  const venv = path.join(os.homedir(), ".virenv/base/bin/python");
  return existsSync(venv) ? venv : "python3";
}

// The default runner: spawn Python, resolve with { code, stdout, stderr }. The
// timeout is generous — cold model load on this CPU is ~2s and a long note a few
// seconds more, but a wedged onnxruntime shouldn't hang a turn forever.
function spawnPython(audioPath, { timeoutMs = 60000 } = {}) {
  return new Promise((resolve, reject) => {
    const proc = spawn(pythonBin(), [TRANSCRIBE_PY, audioPath], {
      env: process.env,
      timeout: timeoutMs,
    });
    let stdout = "";
    let stderr = "";
    proc.stdout.on("data", (d) => (stdout += d));
    proc.stderr.on("data", (d) => (stderr += d));
    proc.on("error", reject);
    proc.on("close", (code) => resolve({ code, stdout, stderr }));
  });
}

/**
 * True if voice transcription can run here: the interpreter exists and it can
 * import onnx_asr and reach ffmpeg. Cheap enough to call at startup; it does not
 * load the model. Mirrors strength/plot.js's plotAvailable().
 */
export function transcriptionAvailable() {
  // A tiny probe process rather than trusting a flag: catches a half-installed
  // venv the same way the real call would, without the model-load cost. It
  // checks the two things the transcriber needs beyond the interpreter —
  // onnx_asr importable and ffmpeg reachable (next to python, or on PATH).
  const probe = spawnSync(
    pythonBin(),
    ["-c", "import onnx_asr, shutil, os, sys; "
      + "sys.exit(0 if (shutil.which('ffmpeg') or "
      + "os.path.isfile(os.path.join(os.path.dirname(sys.executable),'ffmpeg'))) else 1)"],
    { stdio: "ignore" }
  );
  return probe.status === 0;
}

/**
 * Transcribe one audio file to text.
 *
 * @param {string} audioPath        the file to transcribe (Opus/Ogg, wav, …)
 * @param {object} [opts]
 * @param {Function} [opts.run]      injected runner (audioPath, opts) ->
 *                                   Promise<{code, stdout, stderr}>. Defaults to
 *                                   spawning Python; overridden in tests.
 * @param {number} [opts.timeoutMs]
 * @returns {Promise<string>}        the transcript; "" means ran-but-silent.
 * @throws  when the transcriber can't run (missing deps/model, decode error,
 *          timeout) — exit code 2 or a spawn error. The caller distinguishes a
 *          throw ("couldn't transcribe") from "" ("heard nothing").
 */
export async function transcribe(audioPath, { run = spawnPython, timeoutMs } = {}) {
  if (!audioPath) throw new Error("transcribe: no audio path given");
  let result;
  try {
    result = await run(audioPath, { timeoutMs });
  } catch (err) {
    throw new Error(`could not run the transcriber (${pythonBin()}): ${err.message}`);
  }
  const { code, stdout, stderr } = result;
  if (code !== 0) {
    throw new Error(`transcription failed (exit ${code}): ${(stderr || "").trim().slice(-400)}`);
  }
  return (stdout || "").trim();
}
